p2p/forkid.py

```diff
@@ -0,0 +1,116 @@
+import binascii
+import itertools
+import logging
+import sys
+from typing import Tuple, Type
+
+import rlp
+
+from eth.abc import VirtualMachineAPI
+from eth.vm.forks.homestead import HomesteadVM
+
+from eth_typing import BlockNumber, Hash32
+
+from eth_utils import to_bytes
+
+from p2p.exceptions import RemoteChainIsStale, LocalChainIncompatibleOrStale
+
+
+class ForkID(rlp.Serializable):
+    fields = [('hash', rlp.sedes.binary.fixed_length(4)), ('next', rlp.sedes.big_endian_int)]
+
+    def __init__(self, hash: bytes, next: BlockNumber) -> None:
+        # We often instantiate ForkIDs directly (rather than via rlp.decode()), so we validate the
+        # hash length here as well to catch mistakes before we attempt to serialize.
+        if len(hash) != 4:
+            raise ValueError("Hash {hash!r} length is not 4")
+        super().__init__(hash, next)
+
+
+def _extract_fork_blocks(
```

Comment:
I think we usually try to avoid these intermediate lists either by using comprehension syntax or `yield` + `to_tuple` syntax. In this case that would be.
```suggestion
@to_tuple
def _extract_fork_blocks(
        vm_configuration: Tuple[
            Tuple[BlockNumber, Type[VirtualMachineAPI]], ...]
) -> Tuple[BlockNumber, ...]:
    for block, vm in vm_configuration:
        if block == 0:
            continue
        yield block
        if issubclass(vm, HomesteadVM) and vm.support_dao_fork:
            yield vm.get_dao_fork_block_number()
```

Reply:
That's cool!